libGitWrap/Tree.cpp

```diff
@@ -81,19 +81,38 @@ namespace Git
         return PrivatePtr(new Tree::Private(d->repo(), subObject));
     }
 
-    DiffList Tree::diffToTree(Result& result , Tree newTree)
+    /**
+     * @brief           Diff this tree to a newer tree.
+     *
+     * @param[in,out]   result      A Result object; see @ref GitWrapErrorHandling
+     *
+     * @param[in]       newTree     is considered as the "new tree" to compare with
+     *
+     * @return          the differences as @ref DiffList object
+     *
+     * The calling Tree object is considered as the "old tree" and @p newTree
+     * is the "new tree" to compare with. When exchanging both trees, the
+     * result will be inversed.
+     *
+     * An invalid Tree object is is considered an empty tree.
+     */
+    DiffList Tree::diffToTree(Result& result, Tree newTree)
     {
-        GW_D_CHECKED(Tree, DiffList(), result)
+        // d is allowed to be invalid
+        GW_CHECK_RESULT( result, DiffList() );
+        GW_D( Tree );
+
+        Repository::Private* rp = Base::Private::dataOf<Repository>( d ? d->repo() : newTree.repository() );
```

Comment:
This is a gamble for which Repository to use. Actually it should enforce that both sides of the diff come from the same repository. Maybe this is another point, why this style of API isn't the best one.

What about bringing the old semantics back ("this must be valid"), but instead introduce a new static method where any one of the inputed tree parameters is allowed to be invalid?